flare_portal/users/decorators.py

```diff
@@ -31,3 +32,34 @@ def _wrapped_view(
         return _wrapped_view
 
     return decorator(function)
+
+
+def has_researcher_access(
+    function: Callable,
+    owner_only: bool = False,
+    login_url: str = None,
+    redirect_field_name: str = REDIRECT_FIELD_NAME,
+) -> Callable:
+    def decorator(view_func: Callable) -> Callable:
+        @wraps(view_func)
+        def _wrapped_view(
+            request: HttpRequest, *args: Any, **kwargs: Any
+        ) -> HttpResponse:
+            if isinstance(request.user, User):
+                if project_pk := kwargs.get("project_pk"):
+                    # Get all the projects that the user can access.
+                    user_projects = request.user.get_projects(
+                        owner_only=owner_only
+                    ).values_list("id", flat=True)
+
+                    if project_pk in user_projects or request.user.is_admin:
```

Comment:
**Suggestion:** This could be optimised, I think. No need to fetch all the IDs, right?

```python
is_user_project = request.user.get_projects(owner_only=owner_only).filter(pk=project_pk).exists()
```

Also, the second check is cheaper, so I think you should do it first.

```python
if not request.user.is_admin or is_users_project
```

Anyway, I think this kind of flow may look much better:

```python
# acl.py
def can_user_view_project(*, project: Project, user: User) -> bool:
    if not user.is_authenticated:
        return False
    return user.is_admin or user.get_projects(owner_only=owner_only).filter(pk=project).exists()

# view logic
def view():
    project = Project.objects.filter(...).first()
    if not project or not can_user_view_project(project=project, user=user):
        return redirect("experiments:project_list")
    return view_func(...)
```
    
You should treat ACL and HTTP interface as two different things IMO.